processing/src/main/java/io/druid/segment/virtual/ExpressionObjectSelector.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.virtual;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import com.google.common.collect.Maps;
+import com.google.common.primitives.Doubles;
+import io.druid.common.guava.GuavaUtils;
+import io.druid.math.expr.Expr;
+import io.druid.math.expr.Parser;
+import io.druid.segment.ColumnSelectorFactory;
+import io.druid.segment.FloatColumnSelector;
+import io.druid.segment.LongColumnSelector;
+import io.druid.segment.ObjectColumnSelector;
+import io.druid.segment.column.ColumnCapabilities;
+import io.druid.segment.column.ValueType;
+
+import java.util.Map;
+
+public class ExpressionObjectSelector implements ObjectColumnSelector<Number>
+{
+  private final Expr expression;
+  private final Expr.ObjectBinding bindings;
+
+  private ExpressionObjectSelector(Expr.ObjectBinding bindings, Expr expression)
+  {
+    this.bindings = Preconditions.checkNotNull(bindings, "bindings");
+    this.expression = Preconditions.checkNotNull(expression, "expression");
+  }
+
+  public static ExpressionObjectSelector from(ColumnSelectorFactory columnSelectorFactory, Expr expression)
+  {
+    return new ExpressionObjectSelector(createBindings(columnSelectorFactory, expression), expression);
+  }
+
+  private static Expr.ObjectBinding createBindings(ColumnSelectorFactory columnSelectorFactory, Expr expression)
+  {
+    final Map<String, Supplier<Number>> suppliers = Maps.newHashMap();
+    for (String columnName : Parser.findRequiredBindings(expression)) {
+      final ColumnCapabilities columnCapabilities = columnSelectorFactory.getColumnCapabilities(columnName);
+      final ValueType nativeType = columnCapabilities != null ? columnCapabilities.getType() : null;
+      final Supplier<Number> supplier;
+
+      if (nativeType == ValueType.FLOAT) {
+        final FloatColumnSelector selector = columnSelectorFactory.makeFloatColumnSelector(columnName);
```

Comment:
It could but I'm not sure that change is warranted.

Is your concern performance or is it the amount of code in this method? If it's the latter then how about adding a method like `FloatColumnSelector.asSupplier(floatColumnSelector)`?

If it's performance, I guess I'm not _super_ worried about performance in the Expressions related code since it's already doing a lot of sub-optimal stuff anyway (like interpreting expressions on top of boxed numbers). Changing that might involve no longer using Suppliers anyway.